File: games/check_php.py
import re
import os

def check_php_tags(filepath):
    """
    Checks if PHP tags <?php / <?= and ?> are balanced in a file.
    Also detects cases where a closing tag appears before an opening tag on a per-line basis.
    """
    try:
        with open(filepath, 'r', encoding='utf-8') as f:
            content = f.read()
    except Exception as e:
        return [f"ERROR: Could not read file: {e}"]

    issues = []
    
    # 1. Total count check
    opens = re.findall(r'<\?php|<\?=', content)
    closes = re.findall(r'\?>', content)
    
    if len(opens) != len(closes):
        issues.append(f"Mismatched total tags: Opens({len(opens)}) vs Closes({len(closes)})")

    # 2. Sequential balance check (line by line)
    balance = 0
    for i, line in enumerate(content.split('\n'), 1):
        # Ignore tags inside strings or comments (simplistic check)
        # Note: This regex is still basic and might catch tags in strings
        o = len(re.findall(r'<\?php|<\?=', line))
        c = len(re.findall(r'\?>', line))
        
        balance += o - c
        
        if balance < 0:
            issues.append(f"Line {i}: Found closing tag '?>' without matching open tag. Content: {line.strip()}")
            balance = 0 # Reset to continue checking for other errors
            
    # 3. Check for session_start() position
    if "session_start()" in content:
        lines = content.split('\n')
        found_session = False
        for i, line in enumerate(lines, 1):
            if "session_start()" in line:
                # Check if there's any output or non-PHP code before this
                preceding_content = "\n".join(lines[:i-1])
                # Very loose check: if there is any '?>' before session_start, it might be an issue
                if "?>" in preceding_content:
                    issues.append(f"Line {i}: 'session_start()' called after a PHP block was closed. This may cause 'headers already sent' errors.")
                break

    return issues

def main():
    # Use the directory where the script is located
    current_dir = os.path.dirname(os.path.abspath(__file__))
    print(f"--- Scanning directory: {current_dir} ---")
    
    php_files = [f for f in os.listdir(current_dir) if f.endswith('.php')]
    
    if not php_files:
        print("No PHP files found in this directory.")
        return

    total_issues = 0
    files_with_issues = 0

    for file in php_files:
        path = os.path.join(current_dir, file)
        problems = check_php_tags(path)
        
        if problems:
            files_with_issues += 1
            print(f"\n[!] ISSUE FOUND IN: {file}")
            for p in problems:
                print(f"    - {p}")
                total_issues += 1
        else:
            # print(f"[OK] {file}") # Un-comment for verbose output
            pass

    print("\n" + "="*40)
    if total_issues == 0:
        print(f"SUCCESS: All {len(php_files)} PHP files passed the balance check.")
    else:
        print(f"SUMMARY: Found {total_issues} issues across {files_with_issues}/{len(php_files)} files.")
    print("="*40)

if __name__ == "__main__":
    main()

File: games/check_php.php
<?php
/**
 * PHP Maintenance Tool - Tag Balance Checker
 * Scans all .php files in the current directory for unbalanced <?php and ?> tags.
 */

error_reporting(E_ALL);
ini_set('display_errors', 1);

function check_php_tags($filepath) {
    $content = file_get_contents($filepath);
    if ($content === false) {
        return ["ERROR: Could not read file."];
    }

    $issues = [];
    
    // 1. Total count check
    preg_match_all('/<\?php|<\?=/i', $content, $opens);
    preg_match_all('/\?>/', $content, $closes);
    
    $openCount = count($opens[0]);
    $closeCount = count($closes[0]);
    
    if ($openCount !== $closeCount) {
        $issues[] = "Mismatched total tags: Opens($openCount) vs Closes($closeCount)";
    }

    // 2. Sequential balance check (line by line)
    $lines = explode("\n", $content);
    $balance = 0;
    foreach ($lines as $index => $line) {
        $lineNum = $index + 1;
        
        preg_match_all('/<\?php|<\?=/i', $line, $lineOpens);
        preg_match_all('/\?>/', $line, $lineCloses);
        
        $o = count($lineOpens[0]);
        $c = count($lineCloses[0]);
        
        $balance += $o - $c;
        
        if ($balance < 0) {
            $issues[] = "Line $lineNum: Found closing tag '?>' without matching open tag. Content: " . trim($line);
            $balance = 0;
        }
    }
    
    // 3. Check for session_start() position
    if (strpos($content, 'session_start()') !== false) {
        $pos = strpos($content, 'session_start()');
        $preceding = substr($content, 0, $pos);
        
        if (strpos($preceding, '?>') !== false) {
            $lines_before = explode("\n", $preceding);
            $lineNum = count($lines_before);
            $issues[] = "Line $lineNum: 'session_start()' called after a PHP block was closed. This may cause 'headers already sent' errors.";
        }
    }

    return $issues;
}

// Main scan
$dir = __DIR__;
$files = scandir($dir);
$phpFiles = [];
foreach ($files as $f) {
    if (pathinfo($f, PATHINFO_EXTENSION) === 'php' && $f !== basename(__FILE__)) {
        $phpFiles[] = $f;
    }
}

echo "--- Scanning directory: $dir ---\n";
$totalIssues = 0;
$filesWithIssues = 0;

foreach ($phpFiles as $file) {
    $problems = check_php_tags($dir . DIRECTORY_SEPARATOR . $file);
    if (!empty($problems)) {
        $filesWithIssues++;
        echo "\n[!] ISSUE FOUND IN: $file\n";
        foreach ($problems as $p) {
            echo "    - $p\n";
            $totalIssues++;
        }
    }
}

echo "\n" . str_repeat("=", 40) . "\n";
if ($totalIssues === 0) {
    echo "SUCCESS: All " . count($phpFiles) . " PHP files passed the balance check.\n";
} else {
    echo "SUMMARY: Found $totalIssues issues across $filesWithIssues/" . count($phpFiles) . " files.\n";
}
echo str_repeat("=", 40) . "\n";
